nemo/collections/nlp/data/datasets/sgd_dataset/prediction_utils.py

```diff
@@ -21,20 +21,94 @@
 
 
-def get_predicted_dialog_ret_sys_act(dialog, all_predictions, schemas, eval_debug, in_domain_services):
+def carry_over_slots(
+    cur_usr_frame,
+    all_slot_values,
+    slots_relation_list,
+    prev_frame_service,
+    slot_values,
+    sys_prev_slots,
+    last_sys_slots,
+):
+    if prev_frame_service == cur_usr_frame["service"]:
+        return
+    for (service_dest, slot_dest), cands_list in slots_relation_list.items():
+        if service_dest != cur_usr_frame["service"]:
+            continue
+        for service_src, slot_src, freq in cands_list:
+            if freq < MIN_SLOT_RELATION:
+                continue
+            if (
+                service_src == prev_frame_service
+                and service_src in all_slot_values
+                and slot_src in all_slot_values[service_src]
+            ):
+                slot_values[slot_dest] = all_slot_values[service_src][slot_src]
+            if (
+                service_src == prev_frame_service
+                and service_src in sys_prev_slots
+                and slot_src in sys_prev_slots[service_src]
+            ):
+                slot_values[slot_dest] = sys_prev_slots[service_src][slot_src]
+            if (
+                service_src == prev_frame_service
+                and service_src in last_sys_slots
+                and slot_src in last_sys_slots[service_src]
+            ):
+                slot_values[slot_dest] = last_sys_slots[service_src][slot_src]
+
+
+def get_carryover_value(
+    slot,
+    cur_usr_frame,
+    frame_service_prev,
+    all_slot_values,
+    sys_slots_last,
+    sys_slots_agg,
+    slots_relation_list,
+    sys_rets,
+):
+    ext_value = None
+    if slot in sys_slots_agg[cur_usr_frame["service"]]:
+        ext_value = sys_slots_agg[cur_usr_frame["service"]][slot]
+        sys_rets[slot] = ext_value
+    elif (cur_usr_frame["service"], slot) in slots_relation_list:
+        cands_list = slots_relation_list[(cur_usr_frame["service"], slot)]
+        for dmn, slt, freq in cands_list:
+            if freq < MIN_SLOT_RELATION:
+                continue
+            if dmn in all_slot_values and slt in all_slot_values[dmn]:
+                ext_value = all_slot_values[dmn][slt]
+            if dmn in sys_slots_agg and slt in sys_slots_agg[dmn]:
+                ext_value = sys_slots_agg[dmn][slt]
+            if dmn in sys_slots_last and slt in sys_slots_last[dmn]:
+                ext_value = sys_slots_last[dmn][slt]
+    return ext_value
+
+
+def get_predicted_dialog_nemotracker(dialog, all_predictions, schemas, eval_debug, in_domain_services):
     """Update labels in a dialogue based on model predictions.
+    This approach retreives slot values from the history of system actions if slot is active but it can not find it in
+    user utterance overwrite the labels in the turn with the predictions from the model. For test set, these labels are
+    missing from the data and hence they are added.
```

Comment:
they released the labels for test data, this could be removed

Reply:
Done